sparse_ho/criterion.py

```diff
@@ -355,76 +308,69 @@ def __init__(self, X, y, model, sigma, finite_difference_step=None,
-
-    def value(self, mask, dense, mask2, dense2):
-        dof = ((self.X_val[:, mask2] @ dense2 -
-                self.X_val[:, mask] @ dense) @ self.delta)
+    def value_outer_crit(self, X, y, mask, dense, mask2, dense2):
+        dof = ((X[:, mask2] @ dense2 -
+                X[:, mask] @ dense) @ self.delta)
         dof /= self.epsilon
         # compute the value of the sure
-        val = norm(self.y_val - self.X_val[:, mask] @ dense) ** 2
-        val -= self.X_val.shape[0] * self.sigma ** 2
+        val = norm(y - X[:, mask] @ dense) ** 2
+        val -= X.shape[0] * self.sigma ** 2
         val += 2 * self.sigma ** 2 * dof
-
         return val
 
-    def value_test(self, mask, dense):
-        # self.val_test = None
-        val = (
-            norm(self.y_val - self.X_val[:, mask] @ dense) ** 2 /
-            self.X_val.shape[0])
-        self.val_test = val
-        return val
-
-    def compute_rmse(self, mask, dense, beta_star):
-        if beta_star is not None:
-            diff_beta = beta_star.copy()
-            diff_beta[mask] -= dense
-            self.rmse = norm(diff_beta)
-        else:
-            self.rmse = None
-
-    def get_val(self, log_alpha, tol=1e-3):
+    def get_val(self, model, X, y, log_alpha, tol=1e-3):
         # TODO add warm start
         mask, dense, _ = get_beta_jac_iterdiff(
-            self.model.X, self.model.y, log_alpha, self.model,
+            X[self.idx_train], y[self.idx_train], log_alpha, model,
             tol=tol, mask0=self.mask0, dense0=self.dense0, compute_jac=False)
         mask2, dense2, _ = get_beta_jac_iterdiff(
-            self.model.X, self.model.y + self.epsilon * self.delta,
-            log_alpha, self.model,
-            tol=tol, compute_jac=False)
+            X[self.idx_train], y[self.idx_train] + self.epsilon * self.delta,
+            log_alpha, model, tol=tol, compute_jac=False)
 
-        val = self.value(mask, dense, mask2, dense2)
+        val = self.value_outer_crit(mask, dense, mask2, dense2)
 
         return val
 
+    def _init_delta_epsilon(self, X):
+        if self.finite_difference_step:
+            self.epsilon = self.finite_difference_step
+        else:
+            # Use Deledalle et al. 2014 heuristic
+            self.epsilon = 2.0 * self.sigma / (X.shape[0]) ** 0.3
+        rng = check_random_state(self.random_state)
+        self.delta = rng.randn(X.shape[0])  # sample random noise for MCMC step
+        self.init_delta_epsilon = True
+
     def get_val_grad(
-            self, log_alpha, get_beta_jac_v,
+            self, model, X, y, log_alpha, get_beta_jac_v,
             mask0=None, dense0=None, beta_star=None,
             jac0=None, max_iter=1000, tol=1e-3, compute_jac=True):
+        if not self.init_delta_epsilon:
+            self._init_delta_epsilon(X)
+
+        def v(mask, dense):
+            return (2 * X[:, mask].T @ (
+                    X[:, mask] @ dense - y -
```

Comment:
X[:, mask] makes copies of X. If you can avoid calling twice X[:, mask] it will save you time.